packages/hardhat/contracts/periphery/ControllerHelper.sol

```diff
@@ -528,9 +479,139 @@ contract ControllerHelper is UniswapControllerHelper, AaveControllerHelper, IERC
+                            liquidity: decreaseLiquidityParam.liquidity,
+                            liquidityPercentage: decreaseLiquidityParam.liquidityPercentage,
+                            amount0Min: decreaseLiquidityParam.amount0Min,
+                            amount1Min: decreaseLiquidityParam.amount1Min
+                        }),
+                        isWethToken0
+                    );
+                } else if (
+                    data[i].rebalanceVaultNftType == ControllerHelperDataType.RebalanceVaultNftType.MintIntoVault
+                ) {
+                    ControllerHelperDataType.MintIntoVault memory mintIntoVaultParams = abi.decode(
+                        data[i].data,
+                        (ControllerHelperDataType.MintIntoVault)
+                    );
+
+                    ControllerHelperUtil.mintIntoVault(
+                        controller,
+                        vaultId,
+                        mintIntoVaultParams.wPowerPerpToMint,
+                        mintIntoVaultParams.collateralToDeposit
+                    );
+                } else if (
+                    data[i].rebalanceVaultNftType == ControllerHelperDataType.RebalanceVaultNftType.WithdrawFromVault
+                ) {
+                    ControllerHelperDataType.withdrawFromVault memory withdrawFromVaultParams = abi.decode(
+                        data[i].data,
+                        (ControllerHelperDataType.withdrawFromVault)
+                    );
+
+                    ControllerHelperUtil.withdrawFromVault(
+                        controller,
+                        vaultId,
+                        withdrawFromVaultParams.wPowerPerpToBurn,
+                        withdrawFromVaultParams.collateralToWithdraw
+                    );
+                } else if (data[i].rebalanceVaultNftType == ControllerHelperDataType.RebalanceVaultNftType.MintNewLp) {
+                    ControllerHelperDataType.LpWPowerPerpPool memory mintNewLpParams = abi.decode(
+                        data[i].data,
+                        (ControllerHelperDataType.LpWPowerPerpPool)
+                    );
+
+                    uint256 tokenId = ControllerHelperUtil.lpWPowerPerpPool(
+                        controller,
+                        nonfungiblePositionManager,
+                        wPowerPerpPool,
+                        wPowerPerp,
+                        vaultId,
+                        mintNewLpParams
+                    );
+
+                    // deposit Uni NFT token in vault
+                    INonfungiblePositionManager(nonfungiblePositionManager).approve(controller, tokenId);
+                    IController(controller).depositUniPositionToken(vaultId, tokenId);
+                }
+            }
+
+            // remove flashloan amount in ETH from vault + any amount of collateral user want to withdraw (sum <= vault.collateralAmount)
+            IController(controller).withdraw(vaultId, _amount);
```

Comment:
think potential for excess weth, eth and wPowerPerp that need to deal with

Reply:
From where the excess could come ?

Reply:
ETH from withdrawing collateral
WETH from LPing where less weth is needed than expected
wPowerPerp from LPing where less wPowerPerp is needed vs the expected amount minted

Reply:
Any WETH or wPowerPerp excess in LPing or increasing is checked here https://github.com/opynfinance/squeeth-monorepo/blob/feat/rebalance-no-vault/packages/hardhat/contracts/periphery/lib/ControllerHelperUtil.sol#L237

And this at the end should unwrap any WETH excess before sending back https://github.com/opynfinance/squeeth-monorepo/blob/feat/rebalance-no-vault/packages/hardhat/contracts/periphery/lib/ControllerHelperUtil.sol#L253

Reply:
ah I see, its in the LP mint function